### PR TITLE
[Spec decoding] Make spec decoding compatible with attn-DP

### PR DESCRIPTION
**Pull request by gxd3**
[Spec decoding] Make spec decoding compatible with attn-DP

Made fields under `spec_decode_metadata` to be DP-aware layout: [valid entries + padding] * DP-size. This is done in `get_spec_decode_metadata()`

Adjust all downstream consumers of `spec_decode_metadata` accordingly, 
including rejection-sampling; eagle3 propose draft tokens. Most of the changes are wrapping existing functionalities within a jax.shard_map(...)

eagle3/mtp + attn-dp is supported by this PR.
eagle3/mtp + attn-dp + async schleduing need a bit additional changes, will be done in the next PR.

Tested=Added a E2E test that has DP > 1 + eagle3 spec-decoding

### CODE REVIEW

File: tpu_inference/spec_decode/jax/eagle3.py [hunk 1]
```diff
@@ -139,72 +140,85 @@ def _prepare_input_ids(
             num_reqs: jax.Array) -> tuple[jnp.ndarray, jnp.ndarray]:
         """JIT-compiled helper for preparing the input IDs for the draft model."""
 
-        last_token_indices = query_start_loc[1:] - 1
-        # Shift the input ids by one token.
-        rolled_input_ids = jnp.roll(target_token_ids, -1, axis=0)
+        def _body(query_start_loc, target_token_ids, next_token_ids, num_reqs):
```
Comment (Lumosis) [inline tpu_inference/spec_decode/jax/eagle3.py]: What does `body` mean? Maybe clarify it using comments or use a more informative name?

File: tpu_inference/spec_decode/jax/eagle3.py [hunk 2]
```diff
@@ -139,72 +140,85 @@ def _prepare_input_ids(
...
-        clamped_positions = lax.with_sharding_constraint(
-            clamped_positions, NamedSharding(self.mesh, PartitionSpec(None, )))
-        new_seq_lens = lax.with_sharding_constraint(
-            new_seq_lens, NamedSharding(self.mesh, PartitionSpec(None, )))
-        query_start_loc = lax.with_sharding_constraint(
-            query_start_loc, NamedSharding(self.mesh, PartitionSpec()))
-        new_block_tables = lax.with_sharding_constraint(
-            new_block_tables, NamedSharding(self.mesh, PartitionSpec(None, )))
-
+        def _body(positions, seq_lens, block_tables):
```
Comment (Lumosis) [inline tpu_inference/spec_decode/jax/eagle3.py]: Here as well

File: tpu_inference/spec_decode/jax/eagle3.py [hunk 3]
```diff
@@ -378,25 +408,39 @@ def _filter_token_and_prepare_initial_inputs(
...
-        else:
-            input_positions = attn_metadata.input_positions[token_indices]
-
-        attn_metadata = AttentionMetadata(
+        data_spec = PartitionSpec(ShardingAxisName.ATTN_DATA)
+        positions_spec = (PartitionSpec(None, ShardingAxisName.ATTN_DATA)
+                          if attn_metadata.input_positions.ndim == 2 else
+                          data_spec)
+
+        def _select(input_ids, token_indices, input_positions,
```
Comment (Lumosis) [inline tpu_inference/spec_decode/jax/eagle3.py]: Let's use a more informative name?

**Comment by Lumosis** [comment]
Some nit. Could you wait until I enable the MTP tests?

File: tpu_inference/spec_decode/jax/eagle3.py (hunk 1 above)
Comment (gxd3) [reply tpu_inference/spec_decode/jax/eagle3.py]: updated to `_sharded_prepare_input_ids`

File: tpu_inference/spec_decode/jax/eagle3.py (hunk 2 above)
Comment (gxd3) [reply tpu_inference/spec_decode/jax/eagle3.py]: updated to `_sharded_update_inputs_for_loop_speculation`

File: tpu_inference/spec_decode/jax/eagle3.py (hunk 3 above)
Comment (gxd3) [reply tpu_inference/spec_decode/jax/eagle3.py]: updated to `_sharded_select_target_tokens_and_hidden_states`.

**Review by Lumosis — APPROVED** [review]

**Review by wenxindongwork — APPROVED** [review]
